Add unit command bar to TD.

pchote opened the pull request:
This implements the backend for #3789, and the frontend for TD.  The other mods will come in future PRs after this is merged.

![unitcommand](https://cloud.githubusercontent.com/assets/167819/26754716/b9f63b98-4877-11e7-8d6f-023eea379230.png)

The bar exposes clickable UI for all the keyboard-only actions that were previously hacked into `UnitCommandWidget` (now deprecated and will be removed once the bar is implemented in the rest of the mods).

This is important for discoverability and usability / accessibility by players who may not be able to use the keyboard.  For this reason, I consider it to be an essential part of the unit stance/targeting rework that we are aiming for in the next playtest.

pchote committed:
Flip tooltips above the cursor at the bottom of the screen.

pchote committed:
Add IIssueDeployOrder.

pchote committed:
Add Selection.Hash for tracking selection changes.

pchote committed:
Add ForceModifiersOrderGenerator for force move/attack and queueing.

pchote committed:
Add backend code for unit command bar.

cjshmyr commented:
:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

Micr0Bit commented:
>:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

i second that 

:+1:

Micr0Bit reviewed (approved):
>:+1: -- edge scrolling overlaps half of the small buttons but pretty minor

i second that

pchote commented:
Added a commit to fix the edge scrolling, but this then of course means that you lose the ability to edge scroll on the bottom left part of the screen.

I could move the bar up so it is floating with a ~10px gap, like the sidebars, to allow scrolling, but we then lose the benefit of fitt's law for the actions and this forces the chat and debug graphs up too.

pchote committed:
Reduce edge scrolling threshold and polish cursor hotspots.

pchote commented:
Fixed the edge scrolling with a compromise:
* Bar is now floating with a 5px gap between left and bottom edges (bad for fitt's law, but consistent with what C&C3 did).
* Edge scrolling threshold reduced to 5px, and cursor hotspots fixed on TD/RA/TS (D2K was already fine) so that cursors show properly without needing a huge threshold.

@cjshmyr and @Micr0Bit, can you confirm you :+1:s still hold with these changes?

cjshmyr commented:
:+1:

Micr0Bit commented:
:+1:

rob-v commented on OpenRA.Mods.Common/Widgets/CommandBarButtonWidget.cs:
```diff
@@ -0,0 +1,40 @@
+		{
+			GetTooltipDesc = () => TooltipDesc;
+			GetKeyName = () => null;
+		}
```
You could add: OnKeyPress = ki => { if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat) OnClick();
You can compare/hear the difference by holding A in release and this PR.

my suggestion from another PR: change default ButtonWidget.OnKeyPress from { OnClick(); } to {ki => { if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat) OnClick(); }

rob-v reviewed (commented):
Works great, few comments.

It is intentional that you press A key to turn 'Attack move' on once, but you need to hold Alt key, right?
You can turn e.g. Attack move and Force move by commandbar left click, but you can't turn it off with next left click. Left clicks turns on both in the same way, but key press works differently (1 press vs holding).

pchote committed:
Add support for button tooltip descriptions.

pchote committed:
Add ButtonTooltipWithDescHighlightLogic.

Highlights TooltipDesc characters inside {}.

pchote committed:
Add DisableKeyRepeat flag to ButtonWidget.

pchote committed:
Add unit command bar to TD.

pchote commented:
Thanks for the excellent feedback @rob-v.  I have quite significantly reworked the code to take your suggestions, and my own additions based on those, into account.  Needs a rereview.

pchote commented:
Updated screenshot showing the edge scrolling margin and the new force-modifier tooltip style:
![screen shot 2017-06-04 at 14 13 23](https://cloud.githubusercontent.com/assets/167819/26762005/da2af9de-4931-11e7-8ff6-ae274ac30cda.png)

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs:
```diff
@@ -0,0 +1,274 @@
+				deployButton.OnKeyPress = ki =>
+				{
+					if (ki.Event == KeyInputEvent.Down && !ki.IsRepeat)
+						deployButton.OnClick();
```
This is now unnecessary with DisableKeyRepeat?

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs:
```diff
@@ -0,0 +1,274 @@
+
+				deployButton.OnClick = () =>
+				{
+					deployHighlighted = 2;
```
I noticed that there is additonal blick due to this on mouse click. The button is highlighted already on button click. This should be moved to OnKeyPress (also for Scatter, Stop).

rob-v reviewed (commented):
I would feel better if you would keep the modifiers not highlighted due to required changes in new ButtonTooltipWithDescHighlightLogic or at least if it could be unified with also extended ButtonTooltipLogic. You even added highlighting feedback on hotkey press.

Last small issues/questions.

pchote commented:
> or at least if it could be unified with also extended ButtonTooltipLogic. 

Of the three options:
1. Call a virtual method from the base class's constructor
2. Deviate from our standard code patterns for UI logic classes
3. Duplicate ~50 lines of code.

Option 3 is the least bad.

pchote commented:
Fixed the other two issues.

rob-v reviewed (approved):
It is really great. 
👍 (especially for DisableKeyRepeat)

cjshmyr commented:
I've encountered starting unit selections (not control groups) being in a state where there's no command bar options. I've had it happen with:

Starting a game as Nod and selecting all the starting infantry units
Starting a game as GDI and selecting all starting units

![image](https://cloud.githubusercontent.com/assets/566742/26768215/8e322228-4963-11e7-9599-5f3dd880aabd.png)

GraionDilach commented:
Shouldn't this fix #13105?

pchote commented:
> I've encountered starting unit selections being in a state where there's no command bar options.

I couldn't reproduce this (it was a problem in the first version of the PR, but I thought I had fixed it), but just to be safe I replaced the hash with a simple and foolproof counter.

> Shouldn't this fix #13105?

Not this pr, but I will get there in the RA pr.

cjshmyr commented:
Thanks, that worked 👍. @rob-v: anything else?

rob-v commented:
👍

cjshmyr commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/244d86d05a79da2ce6204e5c40c6ddb53e191209...e7618f5d466ed33496075d60ed9f35e82e37e9af), probably want to move that into general improvements once it hits all mods.
